### PR TITLE
Added Prelude/List/unpackOptionals

### PR DESCRIPTION
**Pull request by lisael**
This is useful when the targeted type (say `k8s.Something`) requires plain values, but the data building logic outputs Optionals.

This could probably use fold/build rather than map/concat but I can't wrap my procedural-wired-head around those.

Still WIP. If the need for the feature itself requires more discussion I close this PR and go through a RFC process. Otherwise, and if the implementation is OK, I proceed and add the function in the `package` file and update sha256s accordingly.

### CODE REVIEW

**Comment by sjakobi** [comment]
Thanks! I was surprised that we don't already have this in the Prelude!

I've figured out a slightly simpler definition:

```dhall
\(a : Type) -> Prelude.List.concatMap (Optional a) a (Prelude.Optional.toList a)
```

I think we should try to improve the name a bit too. The Haskell equivalent would be `Data.Maybe.catMaybes`…

`Prelude.Optional.concatList`? `Prelude.List.concatOptionals`?

**Comment by lisael** [comment]
@sjakobi: Thank you for pointing `Optional.toList`, I've updated the PR accordingly

Concerning the naming, I m not happy with `Prelude.List.unpackItems`, myself. Thinking about it, this function belongs `Prelude.Optional`.
I'm not sure about `Prelude.Optional.concatList` because from a user standpoint it's not  a concatenation, it looks like a map + filter. 
`Prelude.Optional.unpackList` if "unpack" is the right term (I think I borrowed "unpack" from [Optional.default doc](https://github.com/dhall-lang/dhall-lang/blob/b0b2e6cb29a478ac731d1bd86b30ec1dc1a21399/Prelude/Optional/default#L2) )

**Comment by Gabriel439** [comment]
Perhaps `Prelude.List.unpackOptionals`, mainly because I think somebody could infer from that name what the type would be

**Comment by joneshf** [comment]
I'll throw out a few more ideas:

* [`compact`](https://pursuit.purescript.org/packages/purescript-filterable/3.0.2/docs/Data.Compactable#v:compact)
* [`filterNotNull`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/filter-not-null.html)
* [`values`](https://package.elm-lang.org/packages/elm-community/maybe-extra/5.2.0/Maybe-Extra#values)

As that last one points out, this is a degenerate case of [`List.filterMap`](https://package.elm-lang.org/packages/elm/core/latest/List#filterMap). That motivates two more ideas:
1. Define `Prelude/List/filterMap` as:
    ```Dhall
    forall (a : Type) ->
    forall (b : Type) ->
    forall (predicate : a -> Optional b) ->
    forall (input : List a) ->
    List b
    ```
   Then this expression becomes:
    ```Dhall
    Prelude/List/filterMap (Optional a) a (\(value : Optional a) -> value)
    ```
1. If there's `Prelude/List/filterMap`, it seems a small jump to call this `Prelude/List/filterOptional` or some such.

Maybe these ideas help the discussion along.

File: Prelude/List/unpackItems [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    = λ(a : Type) → List/concatMap (Optional a) a (Optional/toList a)
+
+let example0 =
+      assert : unpackItems Natural [ Some 1, None Natural, Some 3 ] ≡ [ 1, 3 ]
+
+let example1 =
+        assert
+      : unpackItems Natural ([] : List (Optional Natural)) ≡ ([] : List Natural)
+
+let example1 =
```
Comment (Gabriel439) [inline Prelude/List/unpackItems:24]: Perhaps add these two property tests:

```dhall
\(a : Type) -> \(x : a) -> assert : List/unpackItems a [ Some x ] === [ x ]

\(a : Type) -> assert : List/unpackItems a [ None a ] === ([ ] : List a)
```

**Comment by SiriusStarr** [comment]
I would agree with @joneshf on having `filterMap`/`mapMaybe` (or I guess `mapOptional` in this case).  That's the more general case and probably the more commonly used one, since more often than not I'd wager the list of `Optional`s passed to `unpackItems` would come from mapping some function that outputs an `Optional`.

**Comment by Gabriel439** [comment]
@SiriusStarr: I still think the originally proposed utility would still be necessary to solve the author's use case.  If all we needed was a new higher-order function then `Prelude.List.concatMap` would have already sufficed

**Comment by SiriusStarr** [comment]
> @SiriusStarr: I still think the originally proposed utility would still be necessary to solve the author's use case. If all we needed was a new higher-order function then `Prelude.List.concatMap` would have already sufficed

I didn't particularly mean to suggest that they be mutually exclusive, just that a `filterMap` would be a worthy addition.

**Comment by lisael** [comment]
@Gabriel439 got me with `Prelude.List.unpackOptionals` which gives hints about types, while explaining the purpose of the function itself. `filterOptionals` focus more on definition details than the real purpose of the function IMHO. Anyway, I've got no strong feelings about the name.

I didn't define `List.filterMap` here. I've created #992 for this. It's up to you maintainers to choose to merge this PR first or not.

The tests don't pass, probably on property checks, I'm not sure why. I can typecheck and import locally (using 1.31.1) : https://hydra.dhall-lang.org/build/57283/nixlog/3

Last, thanks you all for the warm welcome :) . The more I dive into dhall, the more I like the project, technically and humanly speaking.

**Comment by sjakobi** [comment]
@lisael To fix CI, you need to update our type-inference test for the Prelude. You can either run `scripts/generate-test-files.sh` or manually add the type pointed out in the CI log to `tests/type-inference/success/preludeB.dhall`.

Regarding the name: I wasn't a big fan of `List/unpackOptionals` initially, but it does indicate that we're eliminating a layer of `Optional`s, so it's not that bad either. ;)

**Comment by lisael** [comment]
```sh
θ79° [lisael:~/src/dhall-lang] unpackItems* ± scripts/generate-test-files.sh
builder for '/nix/store/a455k2ald8l9qw8jbmyylk4gpjprchxi-python2.7-pyOpenSSL-18.0.0.drv' failed with exit code 1; last 10 log lines:
  
  /nix/store/44f3hqpyi391pq9srzkfdzjd9kzh3g4f-python2.7-pyOpenSSL-18.0.0/lib/python2.7/site-packages/OpenSSL/_util.py:54: Error
  ===Flaky Test Report===
  
  test_gmtime_adj_notBefore passed 1 out of the required 1 times. Success!
  test_gmtime_adj_notAfter passed 1 out of the required 1 times. Success!
  test_export_text passed 1 out of the required 1 times. Success!
  
  ===End Flaky Test Report===
  ===================== 3 failed, 493 passed in 9.28 seconds =====================
cannot build derivation '/nix/store/sckx9dxryn3h7v5h3g87sf8s1sp04193-python2.7-urllib3-1.23.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/k5xglp1ws70r21idgiilz789qxkr778s-python2.7-dulwich-0.19.3.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/30wndpj9qcdp91as2p5ngl8zjhb5lrqs-mercurial-4.5.2.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/fkbm6ia4m8ssbiy9f54dzm9nxgdg93xz-nix-prefetch-hg.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/kbr84lyin1g2z6pxv7z82xs13rqzkrim-nix-prefetch-scripts.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/0mrj34k1475q3qyn4kpbda7jbq1pz6qr-cabal2nix-2.9.3.drv': 1 dependencies couldn't be built
cannot build derivation '/nix/store/gyd24b285jvyzvgklw2d4326dc1xpvrh-cabal2nix-dhall.drv': 1 dependencies couldn't be built
[332 built (1 failed), 547 copied (1172.2 MiB), 857.3 MiB DL]
error: build of '/nix/store/gyd24b285jvyzvgklw2d4326dc1xpvrh-cabal2nix-dhall.drv' failed
(use '--show-trace' to show detailed location information)
⏚ θ79° [lisael:~/src/dhall-lang] unpackItems* 5h23m25s 100 ± 
```
I finally edited the file by hand, after 5h23m, a little disapointed :). Yes, my nix store was almost empty. No, I don't feel like I'm chasing the bug in pyOpenSSL package install right now.

**Comment by sjakobi** [comment]
@lisael Did you happen to use the Nix caches documented in [the `dhall-haskell` readme](https://github.com/dhall-lang/dhall-haskell#nix)?

We should probably reproduce them in this repo or add a link, if we haven't yet.

Final bikeshedding attempt regarding the name: `Prelude.Optional.flatten`? `Prelude.List.flattenOptionals`? IMHO "flattening" is the slightly more established word for eliminating a type layer than "unpacking".

**Comment by Gabriel439** [comment]
@lisael: I have a pull request up adding instructions on how to use the scripts without rebuilding things: https://github.com/dhall-lang/dhall-lang/pull/993

**Comment by lisael** [comment]
@sjakobi: Thanks for the cache tip, it did work.
@Gabriel439: Nice, thank you (I figured out what to do with @sjakobi 's link, but that's helpfull too)

Regarding the name, I really have no strong feelings and I have no Haskell nor ML nor functional background. I don't think I have any authority on the terminology, I'll let you decide.

**Comment by Gabriel439** [comment]
@lisael: I added you as a collaborator so that you can merge this yourself once the waiting period is over

**Pull request merged**